agent/util-scripts/pbench-tool-meister

```diff
@@ -0,0 +1,836 @@
+        data-sink will collect and store all the tool data during send
+        operations
+  4. Waits for the message to stop the tool
+  5. Waits for the message to send the tool data remotely
+  6. Repeats steps 3 - 5 until a "terminate" message is received
+
+If a SIGTERM or SIGQUIT signal is sent to the tool meister, any existing
+running tool is shutdown, all local data is removed, and the tool meister
+exits.
+
+A redis [1] instance is used as the communication mechanism between the
+various tool meisters on nodes and the benchmark driver. The redis instance is
+used both to communicate the initial data set describing the tools to use, and
+their parameteres, for each tool meister, as well as a pub/sub for
+coordinating starts and stops of all the tools.
+
+The tool meister is given two arguments when started: the redis server to use,
+and the redis key to fetch its configuration from for its operation.
+
+[1] https://redis.io/
+
+$ sudo dnf install python3-redis
+$ sudo pip3 install python-daemon
+$ sudo pip3 install python-pidfile
+"""
+
+import sys
+import os
+import time
+import json
+import logging
+import tempfile
+import subprocess
+import daemon
+import pidfile
+import redis
+import shutil
+import requests
+import hashlib
+
+from pathlib import Path
+from pbench.server.utils import md5sum
+
+
+class Tool(object):
+    """Encapsulates all the state needed to manage a tool running as a background
+    process.
+
+    The ToolMeister class uses one Tool object per running tool.
+
+    FIXME: this class effecitvely re-implements the former
+    "tool-scripts/base-tool" bash script.
+    """
+
+    def __init__(self, name, group, tool_opts, pbench_bin, tool_dir, logger):
+        self.logger = logger
+        self.name = name
+        self.group = group
+        self.tool_opts = tool_opts
+        self.pbench_bin = pbench_bin
+        self.tool_dir = tool_dir
+        self.screen_name = f"pbench-tool-{group}-{name}"
+        self.start_process = None
+        self.stop_process = None
+        self.post_process = None
+
+    def start(self):
+        """Creates the background `screen` process running the tool's "start"
+        operation.
+        """
+        if (
+            self.start_process is not None
+            or self.stop_process is not None
+            or self.post_process is not None
+        ):
+            self.logger.error(
+                "Tool(%s) has unexpected process already running", self.name
+            )
+            return
+        # screen -dm -L -S \"${screen_name}\" ${pbench_bin}/tool-scripts/${name} --${action} --dir=${tool_output_dir} ${tool_opts[@]}
+        args = [
+            "/usr/bin/screen",
+            "-dmS",
+            self.screen_name,
+            f"{self.pbench_bin}/tool-scripts/{self.name}",
+            "--start",
+            f"--dir={self.tool_dir}",
+            self.tool_opts,
+        ]
+        self.logger.info("%s: start_tool -- %s", self.name, " ".join(args))
```

Comment:
Just a suggestion, Can we add error checking here to see if the screen ran properly?

Reply:
Yes, good catch.

Reply:
Actually, `Popen()` and friend raise exceptions which are caught where a `Tool`'s methods are invoked.